### PR TITLE
MAINT: Make `setxor1d` a bit clearer and speed it up

### PR DESCRIPTION
**Pull request by wufangjie**
We need to find the index which is not the same with the left and right, I think np.logical_and's meaning is more clear and I test this got a speed up

### CODE REVIEW

File: numpy/lib/arraysetops.py [hunk 1]
```diff
@@ -378,8 +378,9 @@ def setxor1d(ar1, ar2, assume_unique=False):
 #    flag = ediff1d( aux, to_end = 1, to_begin = 1 ) == 0
     flag = np.concatenate(([True], aux[1:] != aux[:-1], [True]))
 #    flag2 = ediff1d( flag ) == 0
-    flag2 = flag[1:] == flag[:-1]
-    return aux[flag2]
+#    flag2 = flag[1:] == flag[:-1]
+#    return aux[flag2]
+    return aux[np.logical_and(flag[1:], flag[:-1])]
```
Comment (jaimefrio) [inline numpy/lib/arraysetops.py]: I think they are internally translated to the same calls, but I'd day it's more common to use the bitwise operators for boolean arrays, i.e. do `flag[1:] & flag[:-1]`.

**Review by jaimefrio — CHANGES_REQUESTED** [review]
Thanks, the change LGTM. The existing code was relying on the not obvious property that `flag` can never have two consecutive `False`, so it is indeed much clearer now.

Could you please also:
 * Remove all the commented out code in this function.
 * Add a `MAINT:` prefix to the title of the pull request.

**Comment by eric-wieser** [comment]
This changes the behavior when `assume_unique` is true, but the data isn't. But we probably don't care, and the change might be for the better

**Comment by jaimefrio** [comment]
It does seem to now also be correct if `assume_unique` is wrongly set to `True`, isn't it?

E.g. if the arrays are `[1, 2, 2, 3]` and `[1, 2, 4]`, after concatenating and joining we would get [1, 1, 2, 2, 2, 3, 4] and the flag array would be `[1, 0, 1, 0, 0, 1, 1, 1]` and we will return the correct `[3, 4]` instead of the incorrect `[2, 3, 4]` currently being returned.

If I'm not missing anything, this then begs the question: is it ever really worth making the first two calls to `unique`, now that it works without them? Unless you have huge amounts of repetition it probably never is...

**Comment by eric-wieser** [comment]
You're missing the case where the duplicate is only in one set

**Comment by jaimefrio** [comment]
Ah, right... Too good to be true.

File: numpy/lib/arraysetops.py (hunk 1 above)
Comment (wufangjie) [inline numpy/lib/arraysetops.py]: & is excellent, I forgot np.logical_and is a binary operator sometimes. I'm new to pull request, I did not really understand you two suggestions, it means I should do this next time (you will help me do this time and I don't have to do anything) or I should modify it somewhere, but I can find no way to change the code without a new pull request?

Comment (wufangjie) [inline numpy/lib/arraysetops.py]: & is excellent, I forgot np.logical_and is a binary operator sometimes. Learned a lot, thanks.

**Review by jaimefrio — APPROVED** [review]

**Comment by jaimefrio** [comment]
Seems like you figured out how to make PRs work, in it goes. Thanks @wufangjie!

**Comment by eric-wieser** [comment]
Not a flaw in this process, but seems like `setxor1d([], [])` crashes

**Comment by wufangjie** [comment]
setxor1d([], []) will return 0 since aux.size == 0, a few lines before

**Comment by eric-wieser** [comment]
Right, but flag.shape == (2,), and indexing a (0,) array with a (1,) array is illegal

**Comment by eric-wieser** [comment]
Ah, nevermind

**Comment by charris** [comment]
@wufangjie How would you like your name spelled in the release notes?

**Pull request merged**